Reverting changes from PR #82497

valeriosetti opened the pull request:
As discussed on Discord, PR #82497 introduced an incompatible change with Zephyr's Kconfig pattern (i.e. always enabling `CONFIG_ENTROPY_GENERATOR` even when not requested). A part from the incompatibility, this also caused build issues with some McuBoot samples/tests.

Therefore this PR reverts all the changes introduced by #82497, i.e.:

commit a43cae0b3e40681def6f4a3385fe27e5c0e7ea6c
commit e0fcb8552428fb45164ea2b4271b3b0206695b7f
commit 40e5b8ae77e5ebf16c3f6c7298c85e21f3233685
commit 6f4326ace0febed1f01f935dbcc3761173d79ca1
commit a071257c9b5c394f161e7d4a8ff36f9114b5ef2c
commit 2131b8ba2201ba193345fdf7837682bb48704444
commit 30f6dc43bab8ae1f26497eb000ff1ab4fb379c00
commit 9789d4d5f8cd7fe59dff700f23758857605745fc
commit 5c9689628c2a663d1a275d07db2ead789432edf1

andyross commented:
Hygiene nitpick: if you're going to call something a "revert", revert each patch individually and don't squash them all.  If they were worth splitting out for clarity the first time, then the same aesthetics should apply.

That said: is a kconfig regression really worth a 9-patch reversion? I mean, the workaround for affected apps is just "turn it off", right?  (Either at the kconfig level or by setting it in a DTS override).  Surely there's a way to do a more targeted fix in place?  Basically this doesn't seem like an "emergency rollback!" kind of situation to me.

valeriosetti commented:
> is a kconfig regression really worth a 9-patch reversion?

@andyross 5 of those commits are a follow-up to the initial one (5c9689628c2a663d1a275d07db2ead789432edf1) in order to fix tests issues:
- 2131b8ba2201ba193345fdf7837682bb48704444
- a071257c9b5c394f161e7d4a8ff36f9114b5ef2c
- 6f4326ace0febed1f01f935dbcc3761173d79ca1
- 40e5b8ae77e5ebf16c3f6c7298c85e21f3233685
- e0fcb8552428fb45164ea2b4271b3b0206695b7f

2 are clean up which are only possible if the initial commit is in place:
- 9789d4d5f8cd7fe59dff700f23758857605745fc
- 30f6dc43bab8ae1f26497eb000ff1ab4fb379c00

and 1 is for documentation (a43cae0b3e40681def6f4a3385fe27e5c0e7ea6c) and of course it makes sense only if the initial commit is included.

So yes, I think all of them should be reverted. The clean up ones might come back in the upcoming PR were I try to fix `CONFIG_ENTROPY_GENERATOR` differently, but for the time being they cannot stay because they would break all the fixes I did for BT in #79931.

> Hygiene nitpick: if you're going to call something a "revert", revert each patch individually and don't squash them all. If they were worth splitting out for clarity the first time, then the same aesthetics should apply.

Ok, good to know, I'll do it immediately

valeriosetti committed:
Revert "doc: updates concerning ENTROPY_GENRATOR auto-enabling"

This reverts commit a43cae0b3e40681def6f4a3385fe27e5c0e7ea6c.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
Revert "tests: disable NRF5 entropy driver on some bsim/qemu tests"

This reverts commit e0fcb8552428fb45164ea2b4271b3b0206695b7f.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
Revert "samples: mcumgr: smp_svr: disable "zephyr,entropy" in mcuboot"

This reverts commit 6f4326ace0febed1f01f935dbcc3761173d79ca1.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
Revert "test: misc|shell: disable "zephyr,entropy" chosen property"

This reverts commit 40e5b8ae77e5ebf16c3f6c7298c85e21f3233685.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
Revert "drivers: entropy: nrf5: add "depends on MULTITHREADING""

This reverts commit a071257c9b5c394f161e7d4a8ff36f9114b5ef2c.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
Revert "drivers: entropy: mcux_rnga: add "depends on SYS_CLOCK_EXISTS""

This reverts commit 2131b8ba2201ba193345fdf7837682bb48704444.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
Revert "samples/tests: bluetooth: remove ENTROPY_GENERATOR selection"

This reverts commit 30f6dc43bab8ae1f26497eb000ff1ab4fb379c00.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
Revert "boards: remove ENTROPY_GENERATOR selection if BT"

This reverts commit 9789d4d5f8cd7fe59dff700f23758857605745fc.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
Revert "drivers: entropy: enable ENTROPY_GENERATOR if "zephyr,entropy" is set in DT"

This reverts commit 5c9689628c2a663d1a275d07db2ead789432edf1.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

nordicjm commented on boards/nordic/nrf5340_audio_dk/Kconfig.defconfig:
```diff
@@ -55,6 +55,15 @@ endif # BOARD_NRF5340_AUDIO_DK_NRF5340_CPUAPP_NS
+# instead of manually adding it to all samples/tests configuration files because
+# it looks more compact and easier to maintain.
+config ENTROPY_GENERATOR
+	bool
```
```suggestion
```

nordicjm commented on boards/native/nrf_bsim/Kconfig.defconfig:
```diff
@@ -56,6 +56,15 @@ endif # BOARD_NRF5340BSIM_NRF5340_CPUAPP
+# instead of manually adding it to all samples/tests configuration files because
+# it looks more compact and easier to maintain.
+config ENTROPY_GENERATOR
+	bool
```
```suggestion
```

nordicjm commented on boards/nordic/nrf5340dk/Kconfig.defconfig:
```diff
@@ -45,6 +45,15 @@ endif # BOARD_NRF5340DK_NRF5340_CPUAPP && TRUSTED_EXECUTION_SECURE
+# instead of manually adding it to all samples/tests configuration files because
+# it looks more compact and easier to maintain.
+config ENTROPY_GENERATOR
+	bool
```
```suggestion
```

nordicjm commented on samples/bluetooth/central_ht/boards/frdm_rw612.conf:
```diff
@@ -1 +1,2 @@
 CONFIG_PM=y
+CONFIG_ENTROPY_GENERATOR=y
```
we can't be having this is all these overlays for these boards, should just be done by bluetooth subsystem

nordicjm replied:
ah just a revert PR, missed that from the link

aescolar reviewed (dismissed):
Thanks
But note the CI failure

jhedberg commented:
@valeriosetti reverting the behavior related to `zephyr,entropy` makes sense, but if we now know where the right place to enable the option is, couldn't you leave the app-specific changes unreverted and instead have them rely on the option getting set from the right place?

And do we have consensus on this being the place that's directly making calls which require an entropy driver, i.e. mbedtls/PSA related code? Bluetooth isn't in that case the right place (contrary to what @nordicjm proposed in https://github.com/zephyrproject-rtos/zephyr/pull/82807#discussion_r1879794424) since it's only an indirect user (selecting some mbedtls/PSA options which in turn have the direct dependency).

valeriosetti commented:
> couldn't you leave the app-specific changes unreverted and instead have them rely on the option getting set from the right place?

well, that's an option. Instead of having a PR to deprecate #82497 entirely and then another one which tries to implement it the proper way, I can merge the two (I'm already working on the 2nd one). It depends on how much urgent this fix is: if we need it right now, then revert-all-and-redo is the only option IMO. If I have some more time I can go with a single PR instead

jhedberg commented:
> Bluetooth requires entropy though, at least things like SMP

@cvinayak referenced the Bluetooth Controller, not the Host (which is where e.g. SMP is implemented). I'm not aware of any place in the Bluetooth host code that's directly calling entropy driver APIs or even entropy-related PSA APIs. All that Bluetooth is doing is selecting some PSA/mbedtls Kconfig options, and it should IMO be up to *those* options to know the dependencies they have, either through the appropriate `depends on` or a `select`.

There's a dedicated `bt_rand()` API that the host stack uses everywhere it needs randomness, and that either maps to sending an HCI command to the controller or a HW-specific implementation on the controller side (skipping HCI).

jhedberg commented:
> There's a dedicated `bt_rand()` API that the host stack uses everywhere it needs randomness, and that either maps to sending an HCI command to the controller or a HW-specific implementation on the controller side (skipping HCI).

Sorry, I need to correct myself here. There is actually a third way that `bt_rand()` gets implemented, namely this:
```c
#if defined(CONFIG_BT_HOST_CRYPTO_PRNG)
int bt_rand(void *buf, size_t len)
{
	if (psa_generate_random(buf, len) == PSA_SUCCESS) {
		return 0;
	}

	LOG_ERR("psa_generate_random() failed");
	return -EIO;
}
```

So there the Bluetooth host stack *is* clearly a user of PSA entropy/randomness related APIs. That said, I still feel like we should be getting a Kconfig error out of all this (missing dependency) rather than a link time failure (which is what happens now).

fabiobaltieri commented:
Tagging as hotfix since it fixes https://github.com/zephyrproject-rtos/zephyr/runs/34265176014

fabiobaltieri commented:
Hey folks unfrotunately this does not revert cleanly

```console
$ west build -p -b nrf5340dk/nrf5340/cpuapp/ns samples/net/sockets/http_server -T sample.net.sockets.http.server
...
/usr/local/google/home/fabiobaltieri/zephyrproject/modules/crypto/mbedtls/library/pk.c: In function 'mbedtls_pk_get_psa_attributes':
/usr/local/google/home/fabiobaltieri/zephyrproject/modules/crypto/mbedtls/library/pk.c:475:17: warning: implicit declaration of function 'mbedtls_ecc_group_to_psa'; did you mean 'mbedtls_ecp_group_copy'? [-Wimplicit-function-declaration]
  475 |                 mbedtls_ecc_group_to_psa(ec->grp.id, &bits);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~
      |                 mbedtls_ecp_group_copy
/usr/local/google/home/fabiobaltieri/zephyrproject/modules/crypto/mbedtls/library/pk.c: In function 'copy_from_psa':
/usr/local/google/home/fabiobaltieri/zephyrproject/modules/crypto/mbedtls/library/pk.c:943:18: warning: implicit declaration of function 'mbedtls_ecc_group_from_psa'; did you mean 'mbedtls_pk_copy_from_psa'? [-Wimplicit-function-declaration]
  943 |         grp_id = mbedtls_ecc_group_from_psa(PSA_KEY_TYPE_ECC_GET_FAMILY(key_type), key_bits);
      |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                  mbedtls_pk_copy_from_psa
```

looks like it's missing some option on the test or something? Anyway could you look into it?

valeriosetti commented:
> guess the test depends on it now unfortunately and it's missing some option on the test or something? Anyway could you look into it?

Yes, I'm checking this. Apparently between #82497 and this PR something else was merged that does not allow a clean revert (my guess).

valeriosetti commented:
It seems that this #82143 that was merged yesterday changed the `prj.conf` file of that sample.
Reverting commit https://github.com/zephyrproject-rtos/zephyr/pull//commits/763d9f5bdea9d6c5d3fa8294a9bd7e3c3f11e94b added there solves the build issue but then the build for the nrf5340dk/nrf5340/cpuapp/ns fails due to ROM image not fitting into the Flash memory.

fabiobaltieri commented:
> the build for the nrf5340dk/nrf5340/cpuapp/ns fails due to ROM image not fitting into the Flash memory

if that happens the CI should skip it so at this point I think that would be fine

valeriosetti commented:
> > the build for the nrf5340dk/nrf5340/cpuapp/ns fails due to ROM image not fitting into the Flash memory
> 
> if that happens the CI should skip it so at this point I think that would be fine

For the records, I just checked and 
`west build -p always -b nrf5340dk/nrf5340/cpuapp/ns --test samples/net/sockets/http_server/sample.net.sockets.http.server`
fails with `region FLASH overflowed by 92656 bytes` also at tag `v4.0.0` so this issue is definitely something that has been there for some time now. I'm going to disable the test on this platform.

valeriosetti committed:
samples: sockets: exclude nrf5340dk with TF-M from http_server

When the sample is built for this platform a build error is generated
due to ROM being overflowed.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>
